File: day_12/task_2.py
DIRECTIONS = ["N", "W", "S", "E"]
ROTATIONS = ["L", "R"]

def move_waypoint(position, direction, distance):
  if direction == "N":
    return [position[0], position[1] + distance]
  elif direction == "S":
    return [position[0], position[1] - distance]
  elif direction == "E":
    return [position[0] + distance, position[1]]
  elif direction == "W":
    return [position[0] - distance, position[1]]

def rotate_waypoint(position, direction, rotation):
  if direction == "R":
    for _ in range(int((rotation / 90) % 4)):
      position = [position[1], -position[0]]
  elif direction == "L":
    for _ in range(int((rotation / 90) % 4)):
      position = [-position[1], position[0]]
      
  return position

ship_direction = "E"
ship_position = [0, 0]
waypoint_offset = [10, 1]

file = open("input.txt", "r")

for instruction in file.readlines():
  instruction = instruction.strip("\n")
  command, value = instruction[0], int(instruction[1:])

  if command in DIRECTIONS:
    waypoint_offset = move_waypoint(waypoint_offset, command, value)
  elif command in ROTATIONS:
    waypoint_offset = rotate_waypoint(waypoint_offset, command, value)
  elif command == "F":
    ship_position = [ship_position[0] + (waypoint_offset[0] * value), ship_position[1] + (waypoint_offset[1] * value)]


file.close()

print(sum(map(abs, ship_position)))

File: 2020/day_12/task_2.rb
DIRECTIONS = %w(N W S E)
ROTATIONS = %w(L R)
instructions = File.read("input.txt").split("\n")

def move_waypoint(position, direction, distance)
  case direction
  when "N"
    [position[0], position[1] + distance]
  when "S"
    [position[0], position[1] - distance]
  when "E"
    [position[0] + distance, position[1]]
  when "W"
    [position[0] - distance, position[1]]
  end
end

def rotate_waypoint(position, direction, rotation)
  if direction == "R"
    ((rotation / 90) % 4).times { position = [position[1], -position[0]] }
  elsif direction == "L"
    ((rotation / 90) % 4).times { position = [-position[1], position[0]] }
  end

  position
end

ship_direction = "E"
ship_position = [0, 0]
waypoint_offset = [10, 1]

instructions.each do |instruction|
  command, value = instruction[0], instruction[1..-1].to_i

  if DIRECTIONS.include?(command)
    waypoint_offset = move_waypoint(waypoint_offset, command, value)
  elsif ROTATIONS.include?(command)
    waypoint_offset = rotate_waypoint(waypoint_offset, command, value)
  elsif command == "F"
    ship_position = [ship_position[0] + (waypoint_offset[0] * value), ship_position[1] + (waypoint_offset[1] * value)]
  end
end

p ship_position.map(&:abs).sum
